crates/events/tests/web.rs

```diff
@@ -0,0 +1,191 @@
+    f(Sender { sender });
+
+    receiver
+        .then(|x| match x {
+            Ok(a) => a,
+            Err(_) => unreachable!(),
+        })
+        .collect()
+}
+
+// ----------------------------------------------------------------------------
+// Tests begin here
+// ----------------------------------------------------------------------------
+
+#[wasm_bindgen_test(async)]
+fn new_with_options() -> impl Future<Item = (), Error = JsValue> {
+    mpsc(|sender| {
+        let body = body();
+
+        let _handler = EventListener::new_with_options(
+            &body,
+            "click",
+            &EventListenerOptions {
+                phase: EventListenerPhase::Capture,
+                once: false,
+                passive: false,
+            },
+            move |e| {
+                sender.send(|| {
+                    is(e.dyn_into::<web_sys::MouseEvent>().is_ok(), true)?;
+
+                    Ok(())
+                });
+            },
+        );
+
+        body.click();
+        body.click();
+    })
+    .and_then(|results| {
+        is(results.len(), 2)?;
+        Ok(())
+    })
+}
+
+#[wasm_bindgen_test(async)]
+fn new_with_options_once() -> impl Future<Item = (), Error = JsValue> {
+    mpsc(|sender| {
+        let body = body();
+
+        let _handler = EventListener::new_with_options(
+            &body,
+            "click",
+            &EventListenerOptions {
+                phase: EventListenerPhase::Capture,
+                once: true,
+                passive: false,
+            },
+            move |e| {
+                sender.send(|| {
+                    is(e.dyn_into::<web_sys::MouseEvent>().is_ok(), true)?;
+
+                    Ok(())
+                });
+            },
+        );
+
+        body.click();
+        body.click();
+    })
+    .and_then(|results| {
+        is(results.len(), 1)?;
+        Ok(())
+    })
+}
+
+#[wasm_bindgen_test(async)]
+fn new() -> impl Future<Item = (), Error = JsValue> {
+    mpsc(|sender| {
+        let body = body();
+
+        let _handler = EventListener::new(&body, "click", move |e| {
+            sender.send(|| {
+                is(e.dyn_into::<web_sys::MouseEvent>().is_ok(), true)?;
+
+                Ok(())
+            });
+        });
+
+        body.click();
+        body.click();
+    })
+    .and_then(|results| {
+        is(results.len(), 2)?;
+        Ok(())
+    })
+}
+
+#[wasm_bindgen_test(async)]
+fn once() -> impl Future<Item = (), Error = JsValue> {
+    mpsc(|sender| {
+        let body = body();
+
+        let _handler = EventListener::once(&body, "click", move |e| {
+            sender.send(|| {
+                is(e.dyn_into::<web_sys::MouseEvent>().is_ok(), true)?;
+
+                Ok(())
+            });
+        });
+
+        body.click();
+        body.click();
+    })
+    .and_then(|results| {
+        is(results.len(), 1)?;
+        Ok(())
+    })
+}
+
+// TODO is it possible to somehow cleanup the closure after a timeout?
```

Comment:
We could add `raw` getters for the underlying closure's `js_sys::Function`, and then use `web-sys` manually to remove the listener. This doesn't clean up the Rust closure, however. Probably that's alright (particularly since it isn't actually closing over anything, and therefore should just be a `fn` pointer).

Reply:
> and then use web-sys manually to remove the listener.

Since it's attached to a `<div>` element which doesn't exist in the DOM, the event listeners themself will already be automatically cleaned up.

So the comment was just about the Rust Closure memory.